src/parquet/reader.cc

```diff
@@ -0,0 +1,392 @@
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "parquet/reader.h"
+
+#include <cstdio>
+#include <cstring>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#include "parquet/column_reader.h"
+#include "parquet/exception.h"
+
+#include "parquet/thrift/util.h"
+
+#include "parquet/util/input_stream.h"
+
+using std::string;
+using std::vector;
+using parquet::Type;
+
+namespace parquet_cpp {
+
+// ----------------------------------------------------------------------
+// LocalFile methods
+
+LocalFile::~LocalFile() {
+  CloseFile();
+}
+
+void LocalFile::Open(const std::string& path) {
+  path_ = path;
+  file_ = fopen(path_.c_str(), "r");
+  is_open_ = true;
+}
+
+void LocalFile::Close() {
+  // Pure virtual
+  CloseFile();
+}
+
+void LocalFile::CloseFile() {
+  if (is_open_) {
+    fclose(file_);
+    is_open_ = false;
+  }
+}
+
+size_t LocalFile::Size() {
+  fseek(file_, 0L, SEEK_END);
+  return Tell();
+}
+
+void LocalFile::Seek(size_t pos) {
+  fseek(file_, pos, SEEK_SET);
+}
+
+size_t LocalFile::Tell() {
+  return ftell(file_);
+}
+
+size_t LocalFile::Read(size_t nbytes, uint8_t* buffer) {
+  return fread(buffer, 1, nbytes, file_);
+}
+
+// ----------------------------------------------------------------------
+// RowGroupReader
+
+ColumnReader* RowGroupReader::Column(size_t i) {
+  // TODO: boundschecking
+  auto it = column_readers_.find(i);
+  if (it !=  column_readers_.end()) {
+    // Already have constructed the ColumnReader
+    return it->second.get();
+  }
+
+  const parquet::ColumnChunk& col = row_group_->columns[i];
+
+  size_t col_start = col.meta_data.data_page_offset;
+  if (col.meta_data.__isset.dictionary_page_offset &&
+      col_start > col.meta_data.dictionary_page_offset) {
+    col_start = col.meta_data.dictionary_page_offset;
+  }
+
+  std::unique_ptr<ScopedInMemoryInputStream> input(
+      new ScopedInMemoryInputStream(col.meta_data.total_compressed_size));
+
+  FileLike* source = this->parent_->buffer_;
+
+  source->Seek(col_start);
+
+  // TODO(wesm): Law of demeter violation
+  size_t bytes_read = source->Read(input->size(), input->data());
+
+  if (bytes_read != input->size()) {
+    std::cout << "Bytes needed: " << col.meta_data.total_compressed_size << std::endl;
+    std::cout << "Bytes read: " << bytes_read << std::endl;
+    throw ParquetException("Unable to read column chunk data");
+  }
+
+  // TODO(wesm): This presumes a flat schema
+  ColumnReader* reader = make_column_reader(&col.meta_data,
+      &this->parent_->metadata_.schema[i + 1], input.release());
+
+  column_readers_[i] = std::shared_ptr<ColumnReader>(reader);
```

Comment:
There is a non-zero chance of memory leak if the program crashes before or during line 121. This is marginally better:
column_readers_[i].reset(make_column_reader(...));